File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/remoteconfiguration/ConfigurationBundle.java
package com.snowplowanalytics.snowplow.internal.remoteconfiguration;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.configuration.Configuration;
import com.snowplowanalytics.snowplow.configuration.NetworkConfiguration;
import com.snowplowanalytics.snowplow.configuration.SessionConfiguration;
import com.snowplowanalytics.snowplow.configuration.SubjectConfiguration;
import com.snowplowanalytics.snowplow.configuration.TrackerConfiguration;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class ConfigurationBundle implements Configuration {

    @NonNull
    public final String namespace;

    @Nullable
    public NetworkConfiguration networkConfiguration;
    @Nullable
    public TrackerConfiguration trackerConfiguration;
    @Nullable
    public SubjectConfiguration subjectConfiguration;
    @Nullable
    public SessionConfiguration sessionConfiguration;

    ConfigurationBundle(@NonNull String namespace) {
        this(namespace, null);
    }

    public ConfigurationBundle(@NonNull String namespace, @Nullable NetworkConfiguration networkConfiguration) {
        this.namespace = namespace;
        this.networkConfiguration = networkConfiguration;
    }

    public ConfigurationBundle(@NonNull Context context, @NonNull JSONObject jsonObject) throws JSONException {
        this(jsonObject.getString("namespace"));
        JSONObject json = jsonObject.optJSONObject("networkConfiguration");
        if (json != null) {
            networkConfiguration = new NetworkConfiguration(json);
        }
        json = jsonObject.optJSONObject("trackerConfiguration");
        if (json != null) {
            trackerConfiguration = new TrackerConfiguration(context.getPackageName(), json);
        }
        json = jsonObject.optJSONObject("subjectConfiguration");
        if (json != null) {
            subjectConfiguration = new SubjectConfiguration(json);
        }
        json = jsonObject.optJSONObject("sessionConfiguration");
        if (json != null) {
            sessionConfiguration = new SessionConfiguration(json);
        }
    }

    @NonNull
    public List<Configuration> getConfigurations() {
        List<Configuration> array = new ArrayList<>();
        if (networkConfiguration != null) array.add(networkConfiguration);
        if (trackerConfiguration != null) array.add(trackerConfiguration);
        if (subjectConfiguration != null) array.add(subjectConfiguration);
        if (sessionConfiguration != null) array.add(sessionConfiguration);
        return array;
    }

    // Copyable

    @NonNull
    @Override
    public Configuration copy() {
        ConfigurationBundle copy = new ConfigurationBundle(namespace);
        copy.networkConfiguration = networkConfiguration;
        copy.trackerConfiguration = trackerConfiguration;
        copy.subjectConfiguration = subjectConfiguration;
        copy.sessionConfiguration = sessionConfiguration;
        return copy;
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/ConfigurationBundle.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.configuration

import android.content.Context
import org.json.JSONObject

/**
 * For remote configuration. Bundles [Configuration] objects to apply to a tracker.
 */
class ConfigurationBundle @JvmOverloads constructor(
    val namespace: String,
    networkConfiguration: NetworkConfiguration? = null
) : Configuration {
    var networkConfiguration: NetworkConfiguration? = null
    var trackerConfiguration: TrackerConfiguration? = null
    var subjectConfiguration: SubjectConfiguration? = null
    var sessionConfiguration: SessionConfiguration? = null
    var emitterConfiguration: EmitterConfiguration? = null
    
    val configurations: List<Configuration>
        get() {
            val array: MutableList<Configuration> = ArrayList()
            networkConfiguration?.let { array.add(it) }
            trackerConfiguration?.let { array.add(it) }
            subjectConfiguration?.let { array.add(it) }
            sessionConfiguration?.let { array.add(it) }
            emitterConfiguration?.let { array.add(it) }
            return array
        }

    init {
        this.networkConfiguration = networkConfiguration
    }

    constructor(
        context: Context,
        jsonObject: JSONObject
    ) : this(jsonObject.getString("namespace")) {
        var json = jsonObject.optJSONObject("networkConfiguration")
        json?.let { networkConfiguration = NetworkConfiguration(it) }

        json = jsonObject.optJSONObject("trackerConfiguration")
        json?.let { trackerConfiguration = TrackerConfiguration(context.packageName, it) }

        json = jsonObject.optJSONObject("subjectConfiguration")
        json?.let { subjectConfiguration = SubjectConfiguration(it) }
        
        json = jsonObject.optJSONObject("sessionConfiguration")
        json?.let { sessionConfiguration = SessionConfiguration(it) }

        json = jsonObject.optJSONObject("emitterConfiguration")
        json?.let { emitterConfiguration = EmitterConfiguration(it) }
    }

    fun updateSourceConfig(sourceBundle: ConfigurationBundle) {
        sourceBundle.networkConfiguration?.let {
            if (networkConfiguration == null) networkConfiguration = NetworkConfiguration()
            networkConfiguration?.sourceConfig = it
        }
        sourceBundle.trackerConfiguration?.let {
            if (trackerConfiguration == null) trackerConfiguration = TrackerConfiguration()
            trackerConfiguration?.sourceConfig = it
        }
        sourceBundle.subjectConfiguration?.let {
            if (subjectConfiguration == null) subjectConfiguration = SubjectConfiguration()
            subjectConfiguration?.sourceConfig = it
        }
        sourceBundle.sessionConfiguration?.let {
            if (sessionConfiguration == null) sessionConfiguration = SessionConfiguration()
            sessionConfiguration?.sourceConfig = it
        }
        sourceBundle.emitterConfiguration?.let {
            if (emitterConfiguration == null) emitterConfiguration = EmitterConfiguration()
            emitterConfiguration?.sourceConfig = it
        }
    }

    // Copyable
    override fun copy(): Configuration {
        val copy = ConfigurationBundle(namespace)
        copy.networkConfiguration = networkConfiguration
        copy.trackerConfiguration = trackerConfiguration
        copy.subjectConfiguration = subjectConfiguration
        copy.sessionConfiguration = sessionConfiguration
        copy.emitterConfiguration = emitterConfiguration
        return copy
    }
}
